### PR TITLE
[main] Restructure build pipeline

### PR DESCRIPTION
**Pull request by RussKie**
Test build: https://dev.azure.com/dnceng/internal/_build/results?buildId=2406682

 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/2832)

### CODE REVIEW

File: eng/pipelines/azure-pipelines.yml [hunk 1]
```diff
@@ -63,169 +91,137 @@ resources:
...
-        preSteps:
-          - checkout: self
-            fetchDepth: 1
-            clean: true
-
-        steps:
-        - script: $(Build.SourcesDirectory)/build.sh --ci --restore
-          displayName: Init toolset
-
-        - template: /eng/pipelines/templates/VerifyCoverageReport.yml
```
Comment (RussKie) [inline eng/pipelines/azure-pipelines.yml:211]: @davidfowl @eerhardt FYI we're currently dropping the CC report from the internal pipeline. Cobertura files are still being collected and published as artifacts, however, those won't get aggregated or turned into an HTML report. 
This functionality we can bring back later (if required/subject to prioritisation).

Comment (eerhardt) [reply eng/pipelines/azure-pipelines.yml:211]: Honestly, I'm not sure we are even using/looking at code coverage from our PR validation builds. At least I haven't been.

**Review by joperezr — APPROVED** [review]
LGTM other than the comments I left. Final one: shouldn't azure-pipelines.yml be renamed to azure-pipelines-official.yml?

**Comment by RussKie** [comment]
> shouldn't azure-pipelines.yml be renamed to azure-pipelines-official.yml?

No.

File: eng/pipelines/templates/BuildAndTest.yml [hunk 2]
```diff
@@ -43,18 +45,10 @@ steps:
   - ${{ if ne(parameters.skipTests, 'true') }}:
     - script: ${{ parameters.dotnetScript }} dotnet-coverage collect
               --settings $(Build.SourcesDirectory)/eng/CodeCoverage.config
-              --output ${{ parameters.repoTestResultsPath }}/NonHelixTests.cobertura.xml
+              --output ${{ parameters.repoTestResultsPath }}/NonHelixCodeCoverage/NonHelix.cobertura.xml
```
Comment (joperezr) [inline eng/pipelines/templates/BuildAndTest.yml:48]: Should repoTestResultsPath here also change to repoArtifactsPath?

File: eng/pipelines/templates/BuildAndTest.yml [hunk 3]
```diff
@@ -89,23 +85,25 @@ steps:
...
+          publishLocation: 'pipeline'
 
-    - task: PublishBuildArtifacts@1
-      inputs:
-        PathtoPublish: '${{ parameters.repoTestResultsPath }}/$(Agent.Os)_$(Agent.JobName).cobertura.xml'
-        PublishLocation: Container
-        ArtifactName: CodeCoverageResults
-      displayName: Publish coverage results (cobertura.xml)
+    - ${{ if ne(parameters.runAsPublic, 'true') }}:
+      - task: 1ES.PublishPipelineArtifact@1
```
Comment (joperezr) [inline eng/pipelines/templates/BuildAndTest.yml:102]: What is the difference between PublishBuildArtifact and PublishPipelineArtifact?

File: eng/pipelines/azure-pipelines.yml [hunk 4]
```diff
@@ -5,6 +5,20 @@ trigger:
     - main
     - release/*
     - internal/release/*
+  paths:
```
Comment (joperezr) [inline eng/pipelines/azure-pipelines.yml:8]: Why are we adding these exclusions? These don't seem related to the migration.

File: eng/pipelines/templates/BuildAndTest.yml (hunk 2 above)
Comment (RussKie) [reply eng/pipelines/templates/BuildAndTest.yml:48]: Technically we could, we won't gain anything.

File: eng/pipelines/templates/BuildAndTest.yml (hunk 3 above)
Comment (RussKie) [reply eng/pipelines/templates/BuildAndTest.yml:102]: I used it as per a suggestion in the Teams channel.

In summary, if you need to share build outputs within the same pipeline or across different pipelines, use the “Publish Build Artifact” task. [If you want a more efficient way to transfer files between jobs within the same pipeline, opt for the “Publish Pipeline Artifact” task](https://learn.microsoft.com/en-us/azure/devops/pipelines/artifacts/pipeline-artifacts?view=azure-devops).

File: eng/pipelines/azure-pipelines.yml (hunk 4 above)
Comment (RussKie) [reply eng/pipelines/azure-pipelines.yml:8]: Indeed, it's not directly related to the migration. It's an optimisation I copied from dotnet/extensions pipelines.
